File: web/lib/markets.ts
/** UI market views (cn / en). Source market is cn / en / global (global → EN). */

export type MarketId = "cn" | "en";
export type SourceMarket = "cn" | "en" | "global";

export interface MarketConfig {
  id: MarketId;
  label: string;
  defaultCountry: string | null;
  filters: readonly string[];
  defaultCollectSources: readonly string[];
  route: string;
}

export const MARKET_ORDER: MarketId[] = ["cn", "en"];

export const MARKETS: Record<MarketId, MarketConfig> = {
  cn: {
    id: "cn",
    label: "CN",
    defaultCountry: "CN",
    filters: ["source"],
    defaultCollectSources: ["zhaopin", "liepin"],
    route: "/cn",
  },
  en: {
    id: "en",
    label: "EN",
    defaultCountry: null,
    filters: ["country", "remote", "posted", "source", "sponsorship_display"],
    defaultCollectSources: ["linkedin", "hiring_cafe"],
    route: "/en",
  },
};

export const EN_COUNTRY_SEED: { code: string; name: string }[] = [
  { code: "GB", name: "United Kingdom" },
  { code: "NL", name: "Netherlands" },
  { code: "US", name: "United States" },
  { code: "CA", name: "Canada" },
  { code: "AU", name: "Australia" },
  { code: "NZ", name: "New Zealand" },
  { code: "JP", name: "Japan" },
];

export const UNKNOWN_COUNTRY = { code: "XX", name: "Unknown / Global" };

export function parseSourceMarket(raw: string | undefined | null): SourceMarket | null {
  if (!raw) return null;
  const lower = raw.trim().toLowerCase();
  if (lower === "cn" || lower === "en" || lower === "global") return lower;
  return null;
}

export function parseMarketId(raw: string | undefined | null): MarketId | null {
  const sm = parseSourceMarket(raw);
  if (sm === "global") return "en";
  if (sm === "cn" || sm === "en") return sm;
  return null;
}

export function marketFromPath(pathname: string): MarketId | null {
  const first = pathname.split("/").filter(Boolean)[0];
  return parseMarketId(first ?? "");
}

export function marketHasFilter(market: MarketId, name: string): boolean {
  return MARKETS[market].filters.includes(name);
}

export function storedToMarketId(stored: string | undefined | null): MarketId | null {
  return parseMarketId(stored);
}

export function sourceInMarket(sourceMarket: string | undefined | null, market: MarketId): boolean {
  const sm = parseSourceMarket(sourceMarket);
  if (sm == null) return false;
  if (sm === "global") return market === "en";
  return sm === market;
}

export function jobInMarketView(
  job: { source?: string; market?: string; country?: string },
  view: MarketId,
  sourceMarketById: Record<string, string | undefined>,
): boolean {
  const fromRegistry = sourceMarketById[(job.source || "").toLowerCase()];
  const sm = parseSourceMarket(fromRegistry) ?? parseSourceMarket(job.market);
  if (sm == null) return false;
  if (sm === "cn") return view === "cn";
  return view === "en";
}

export function countryFilterOptions(
  jobs: readonly { country?: string; country_name?: string }[],
): { code: string; name: string }[] {
  const byCode = new Map<string, string>();
  for (const row of EN_COUNTRY_SEED) byCode.set(row.code, row.name);
  for (const job of jobs) {
    const code = (job.country || "XX").toUpperCase();
    if (code === "UK") {
      byCode.set("GB", "United Kingdom");
      continue;
    }
    if (code === "XX" || code === "UNKNOWN") continue;
    if (!byCode.has(code)) byCode.set(code, job.country_name || code);
  }
  const rows = [...byCode.entries()]
    .map(([code, name]) => ({ code, name }))
    .sort((a, b) => a.name.localeCompare(b.name));
  rows.push(UNKNOWN_COUNTRY);
  return rows;
}

export function jobsPath(market: MarketId): string {
  return `/${market}/jobs`;
}

export function myJobsPath(_market: MarketId): string {
  return "/tasks";
}

export function searchPath(market: MarketId): string {
  return `/${market}/search`;
}

File: src/job_sentinel/markets.py
"""Market views (cn / en) vs source_market (cn / en / global).

source_market is configured on each collect source. country is a per-job
location attribute. CN and EN views do not mix sources. global boards belong
to the EN view. Missing source_market is never guessed.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Literal

if TYPE_CHECKING:
    from job_sentinel.core.models import Job

MarketId = Literal["cn", "en"]
SourceMarket = Literal["cn", "en", "global"]

_SOURCE_ALIASES: dict[str, SourceMarket] = {
    "cn": "cn",
    "en": "en",
    "global": "global",
}


@dataclass(frozen=True)
class MarketConfig:
    id: MarketId
    label: str
    default_country: str | None
    filters: frozenset[str]
    default_collect_sources: tuple[str, ...]
    route: str


MARKETS: dict[MarketId, MarketConfig] = {
    "cn": MarketConfig(
        id="cn",
        label="CN",
        default_country="CN",
        # CN jobs may leave visa/sponsorship empty; no sponsorship_display filter.
        filters=frozenset({"source"}),
        default_collect_sources=("zhaopin", "liepin"),
        route="/cn",
    ),
    "en": MarketConfig(
        id="en",
        label="EN",
        default_country=None,
        # EN view includes global boards; fill visa/work-permit when available.
        filters=frozenset({"country", "remote", "posted", "source", "sponsorship_display"}),
        default_collect_sources=("linkedin", "hiring_cafe"),
        route="/en",
    ),
}


def parse_source_market(raw: str | None) -> SourceMarket | None:
    """Canonical source_market: cn, en, or global. None if missing/unknown."""
    if raw is None:
        return None
    text = raw.strip()
    if not text:
        return None
    return _SOURCE_ALIASES.get(text.lower())


def require_source_market(raw: str | None) -> SourceMarket:
    parsed = parse_source_market(raw)
    if parsed is None:
        raise ValueError("source market must be cn, en, or global")
    return parsed


def parse_market_id(raw: str | None) -> MarketId | None:
    """UI view id. global international boards map onto EN."""
    sm = parse_source_market(raw)
    if sm == "global":
        return "en"
    if sm == "cn" or sm == "en":
        return sm
    return None


def require_market_id(raw: str | None) -> MarketId:
    mid = parse_market_id(raw)
    if mid is None:
        raise ValueError("Unknown market")
    return mid


def stored_values_for(market_id: MarketId) -> frozenset[str]:
    if market_id == "cn":
        return frozenset({"cn", "CN"})
    return frozenset({"en", "EN", "global", "GLOBAL"})


def market_has_filter(market_id: MarketId, name: str) -> bool:
    return name in MARKETS[market_id].filters


def source_in_view(source_market: str | None, view: MarketId) -> bool:
    """CN sources stay in CN. EN and global boards stay in EN."""
    sm = parse_source_market(source_market)
    if sm is None:
        return False
    if sm == "global":
        return view == "en"
    return sm == view


def resolve_job_source_market(
    *,
    source_id: str,
    stored_market: str,
    registry: dict[str, SourceMarket],
) -> SourceMarket | None:
    """Prefer the source registry, then the stored job market. Never guess."""
    key = source_id.strip().lower()
    if key in registry:
        return registry[key]
    return parse_source_market(stored_market)


def job_in_view(
    *,
    source_market: SourceMarket | None,
    country: str,
    view: MarketId,
) -> bool:
    """Market membership is by source. Country is an EN filter only."""
    del country
    if source_market is None:
        return False
    if source_market == "cn":
        return view == "cn"
    return view == "en"


def job_in_market_view(
    job: Job,
    view: MarketId,
    registry: dict[str, SourceMarket],
) -> bool:
    sm = resolve_job_source_market(
        source_id=job.source,
        stored_market=job.market,
        registry=registry,
    )
    return job_in_view(source_market=sm, country=job.country, view=view)
